src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp

```diff
@@ -1909,6 +1920,13 @@ bool ReplicatedMergeTreeMergePredicate::canMergeTwoParts(
 
     for (const MergeTreeData::DataPartPtr & part : {left, right})
     {
+        if (pinned_part_uuids.part_uuids.contains(part->uuid))
```

Comment:
I think we should not only refuse to execute merges of pinned parts but also don't assign such merges.

Reply:
This method is called from mergeSelectingTask which is the method assigning merges. We don't assign merges for pinned parts.